packages/tpetra/core/src/Tpetra_Details_extractBlockDiagonal.hpp

```diff
@@ -0,0 +1,106 @@
+// @HEADER
+// ***********************************************************************
+//
+//          Tpetra: Templated Linear Algebra Services Package
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+#define TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+
+#include "TpetraCore_config.h"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Teuchos_RCP.hpp"
+
+/// \file Tpetra_Details_extractBlockDiagonal.hpp
+/// \brief Functions that allow for the extraction of a block diagonal
+/// from a Tpetra::CrsMatrix.
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra.  The file itself or its contents may disappear or
+///   change at any time.
+
+namespace Tpetra {
+namespace Details {
+
+
+template<class SparseMatrixType,
+         class MultiVectorType>
+void extractBlockDiagonal(const SparseMatrixType& A, MultiVectorType & diagonal) {
```

Comment:
Do you really want to extract into a MultiVector?  It seems like the lowest-level interface would extract into a rank-3 `Kokkos::View`, where the leftmost index is the local block row index.

Reply:
> It looks like this is actually a band extraction routine. Perhaps we should call it that instead?

Which is isn't supposed to be.  Good catch!